experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,957 @@
+  uint64_t result = 0;
+  uint64_t tmp;
+  for (unsigned i = 0; i < bitwidth; ++i) {
+    tmp = v[i];
+    result |= tmp << (bitwidth - i - 1);
+  }
+  return result;
+}
+
+/// convert integer to vector<bool>
+std::vector<bool> num2vec(uint64_t num, unsigned bitwidth) {
+  std::vector<bool> tempVec;
+  while (num) {
+    tempVec.push_back(num & 1u);
+    num >>= 1;
+  }
+  if (bitwidth > tempVec.size()) {
+    for (unsigned i = tempVec.size(); i < bitwidth; ++i) {
+      tempVec.push_back(false);
+    }
+  } else {
+    for (unsigned i = bitwidth; i < tempVec.size(); ++i) {
+      tempVec.pop_back();
+    }
+  }
+  std::reverse(tempVec.begin(), tempVec.end());
+  return tempVec;
+}
+
+/// represents the signal
+/// in terms of VHDL : std_logic
+class Signal : public Wire {
+  std::optional<bool> signal;
+
+public:
+  Signal(std::optional<bool> s = {}) : signal(s) {}
+  bool hasValue() override { return signal.has_value(); }
+  void updateValue(std::any val) override { signal = std::any_cast<bool>(val); }
+  void printValue() override {
+    if (hasValue()) {
+      llvm::outs() << signal.value();
+    } else {
+      llvm::outs() << "U";
+    }
+  }
+  std::any getData() override { return signal.value(); }
+  virtual ~Signal(){};
+};
+
+/// represents the data
+/// in terms of VHDL : std_logic_vector(WIDTH - 1 downto 0);
+class Data : public Wire {
+  std::optional<std::vector<bool>> data;
+
+public:
+  Data(const std::optional<std::vector<bool>> &d = {}) : data(d) {}
+  void updateValue(std::any val) override {
+    auto vec = std::any_cast<std::vector<bool>>(val);
+    if (data.has_value()) {
+      data->clear();
+      data->insert(data->begin(), vec.begin(), vec.end());
+    } else {
+      data = std::make_optional(vec);
+    }
+  }
+  bool hasValue() override { return data.has_value(); }
+  std::any getData() override { return data.value(); }
+  void printValue() override {
+    if (hasValue()) {
+      std::vector<bool> tempVec = data.value();
+      llvm::outs() << tempVec.size() << " : "
+                   << vec2num(tempVec, tempVec.size()) << "=";
+      for (bool c : tempVec) {
+        llvm::outs() << c;
+      }
+    } else {
+      llvm::outs() << "U";
+    }
+  }
+  virtual ~Data(){};
+};
+
+/// Represents a container for the set of channels inside the CIRCUIT
+/// Weak pointers - because we don't want to change any port from the circuit
+/// itself
+class Channel {
+public:
+  std::weak_ptr<Wire> valid;
+  std::weak_ptr<Wire> ready;
+  std::weak_ptr<Wire> data;
+};
+
+/// Represents a container for the set of channels inside the COMPONENT
+/// Shared pointers - because components can change ports
+class PortChannel {
+public:
+  Port valid;
+  Port ready;
+  Port data;
+};
+
+/// Abstract class to represent any component == execution model
+class Component {
+public:
+  // execute (run along the architecture).
+  // sig == 0 => rst; sig == 1 => clk
+  virtual void exec(bool sig) = 0;
```

Comment:
It looks like this cannot model an "execution" that is just combinationally propagating signals within a cycle (when reset is not active and the clock is not on the rising edge), which is a big problem. I am also not convinced that we should model reset in this execution function at all, although that can be discussed.